theano/gpuarray/opt.py

```diff
@@ -751,16 +751,12 @@ def local_gpua_elemwise(op, context_name, inputs, outputs):
         gpu_output = res(*new_inputs)
         return [gpu_output]
     elif op.scalar_op in (scalar.add, scalar.mul):
-        max_nb_inputs = max_inputs_to_GpuElemwise(outputs)
-        if max_nb_inputs > 1:
-            while len(inputs) > max_nb_inputs:
-                inputs = inputs[:-max_nb_inputs] + [res(*inputs[-max_nb_inputs:])]
-        return res(*inputs)
+        return split_huge_add_or_mul(outputs[0].owner, res).outputs
```

Comment:
I see, thanks for the clarification!

Let's decouple a few things then. I propose this for the optimization.
```python
try:
    return [split_inputs(inputs, max_inputs_to_GpuElemwise(outputs), res)]
except ValueError:
    return False
```

And a bit more general function to split the input.
```python
def split_inputs(inputs, max_nb_inputs, op):
    if max_nb_inputs <= 1 and len(node.inputs) > 1:
        ValueError("Can not split nodes because inputs' dimensionality and/or number of outputs is too large")

    while len(inputs) > max_nb_inputs:
        inner_ops = []
        for i in range(0, len(node.inputs), max_nb_inputs):
            inner_ops.append(op(*inputs[i: i + max_nb_inputs]))
        inputs = inner_ops

    return op(*inputs)
```

The try-except clause combined with raise ValueError will adress this [comment](#r112282662).